src/main/java/fi/aalto/cs/apluscourses/ui/ideactivities/PopupTransparencyHandler.java

```diff
@@ -0,0 +1,65 @@
+package fi.aalto.cs.apluscourses.ui.ideactivities;
+
+public class PopupTransparencyHandler {
+
+  private final TransparentComponent component;
+
+  private long lastTransitionUpdate;
+
+  private static final float MIN_TRANSPARENCY = 0.55f;
+  private static final float MAX_TRANSPARENCY = 1.0f;
+  private static final float TRANSITION_STEP = 0.0015f;
+
+  public PopupTransparencyHandler(TransparentComponent component) {
+    this.component = component;
+  }
+
+  public boolean isInAnimation() {
+    var coefficient = component.getTransparencyCoefficient();
+    return !(coefficient == MIN_TRANSPARENCY || coefficient == MAX_TRANSPARENCY);
+  }
+
+  public void resetAnimationProgress() {
+    lastTransitionUpdate = 0;
+  }
+
+  /**
+   * Updates the status of transparency fading animation.
+   *
+   * @param isMouseOnPopup If true, the mouse cursor is currently on the popup.
+   */
+  public boolean update(boolean isMouseOnPopup) {
+    if (lastTransitionUpdate == 0) {
+      lastTransitionUpdate = System.currentTimeMillis();
+      return true;
+    }
+
+    float transitionStep = isMouseOnPopup ? TRANSITION_STEP : -TRANSITION_STEP;
+
+    long timeSinceLastUpdate = System.currentTimeMillis() - lastTransitionUpdate;
+    float currentCoefficient = component.getTransparencyCoefficient();
+
+    boolean shouldRedraw = true;
+
+    currentCoefficient += transitionStep * timeSinceLastUpdate;
+
+    if (currentCoefficient < MIN_TRANSPARENCY) {
+      currentCoefficient = MIN_TRANSPARENCY;
+      resetAnimationProgress();
+
+      shouldRedraw = false;
+    }
+
+    if (currentCoefficient > MAX_TRANSPARENCY) {
+      currentCoefficient = MAX_TRANSPARENCY;
+      resetAnimationProgress();
+
+      shouldRedraw = false;
```

Comment:
Would omitting redraw cause the popup to never achieve min/max levels? Anyway, if this animation redraws the component, what, ~10 times maybe, does it matter if there's one time more? Or did I get something wrong?

Reply:
You are somehow right, however the bottom line is that `shouldRedraw` is now useless. It was introduced to only redraw the component during animation, but has since been replaced by `isInAnimation`.

We need to redraw the component while the animation is running because this makes the animation look smooth. We also cannot just redraw the component every frame ("GUI update") because it'll consume 9000% of the CPU cycles.